### PR TITLE
Allow OS specific gold files

### LINKED ISSUE #1900: [TASK] Specify OS dependent gold files
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
Some calculations (ARMA for example) give slightly different results when run on Windows, Mac, or Linux. It would be nice if we can specify gold files for each OS in the `tests` files.

**Describe the solution you'd like**
Specify the gold file you would like to use for the test based on the OS. For example: 
```
[./testBasic]
  type = 'RavenFramework'
  input = 'test_basic.xml'
  UnorderedCsv = 'basic.csv'
  windows_gold = 'winBasic.csv'
  mac_gold = 'macBasic.csv'
  linux_gold = 'linuxBasic.csv'
[./] 
```

**Describe alternatives you've considered**
Writing separate tests for each OS and using the "skip_if_OS" parameter.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by dgarrett622**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Addresses #1900 
close #1900 

##### What are the significant changes in functionality due to this change request?
For testing, OS specific gold files may be specified using parameters `windows_gold`, `mac_gold`, or `linux_gold` as seen in the test file `tests/framework/PostProcessors/LimitSurface/tests` under `./testLimitSurfaceIntegralPPWithBoundingError`

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: rook/Tester.py [hunk 1]
```diff
@@ -221,12 +224,28 @@ def _get_gold_files(self):
     """
       returns a list of the full path to the gold files
       @ In, None
-      @ Out, _get_gold_files, List(Strings), the path of the gold files.
+      @ Out, paths, List(Strings), the paths of the gold files.
     """
-    if len(self.specs['gold_files']) > 0:
+    this_OS = platform.system().lower()
+    available_OS = ['windows', 'mac', 'linux'] # list of OS with specific gold file options
```
Comment (dylanjm) [inline rook/Tester.py:230]: I'm assuming rook doesn't have to follow RAVEN's camelCase requirements?

**Review by dylanjm — APPROVED** [review]
Changes appear good to me. I wonder if there is a way that we can just name the gold file we want to compare to instead of having three separate input parameters?

File: rook/Tester.py (hunk 1 above)
Comment (dgarrett622) [reply rook/Tester.py:230]: I think so. I was trying to match the coding style I saw in the existing code.

**Comment by dylanjm** [comment]
I'm not sure where the civet testing suite interface went. @joshua-cogliati-inl does this have something to do with how CIVET was down earlier this week?

**Comment by dgarrett622** [comment]
> Changes appear good to me. I wonder if there is a way that we can just name the gold file we want to compare to instead of having three separate input parameters?

There is a "gold_files" input parameter where you can name the gold file you want to compare to, but if the results are OS dependent, there isn't currently a way to specify them. I suppose if a naming convention for gold files were to be imposed that included OS information in the name of the gold file, this would solve that issue and a different change would be necessary.

**Comment by dylanjm** [comment]
> There is a "gold_files" input parameter where you can name the gold file you want to compare to, but if the results are OS dependent, there isn't currently a way to specify them. I suppose if a naming convention for gold files were to be imposed that included OS information in the name of the gold file, this would solve that issue and a different change would be necessary.

Ah okay, that makes sense to me. This seems like best solution then.

**Comment by dylanjm** [comment]
@dgarrett622 Looks like we don't allow snake case get past the pylint check despite the whole rook codebase using snake case...🤦

**Comment by dgarrett622** [comment]
@dylanjm I think I mixed upper case and lower case in the variable names, which doesn't seem to be allowed with snake_case. I switched it to all lower case and split up some lines, maybe that will work 🤷‍♂️

File: rook/Tester.py [hunk 2]
```diff
@@ -173,7 +173,12 @@ def get_valid_params():
     params = _ValidParameters()
     params.add_required_param('type', 'The type of this differ')
     params.add_required_param('output', 'Output files to check')
-    params.add_param('gold_files', '', 'Gold filenames')
+    params.add_param('windows_gold', '', 'Paths to Windows specific gold files,'+
+                     ' relative to test directory')
+    params.add_param('mac_gold', '', 'Paths to Mac specific gold files, relative to test directory')
+    params.add_param('linux_gold', '', 'Paths to Linux specific gold files,'+
+                     ' relative to test directory')
+    params.add_param('gold_files', '', 'Paths to gold files, relative to test directory')
```
Comment (wangcj05) [inline rook/Tester.py:181]: I guess our assumption here is to use the path relative to gold directory

File: rook/Tester.py [hunk 3]
```diff
@@ -221,12 +226,28 @@ def _get_gold_files(self):
...
+    if (this_os in available_os) and (len(self.specs[f'{this_os}_gold']) > 0):
+      gold_files = self.specs[f'{this_os}_gold'].split()
+      paths = [os.path.join(self.__test_dir, f) for f in gold_files]
+    # if OS specific gold files are not given, are specific gold files given?
+    elif len(self.specs['gold_files']) > 0:
       gold_files = self.specs['gold_files'].split()
-      return [os.path.join(self.__test_dir, f) for f in gold_files]
-    return [os.path.join(self.__test_dir, "gold", f) for f in self.__output_files]
+      paths = [os.path.join(self.__test_dir, f) for f in gold_files]
+    # otherwise, use output files
```
Comment (wangcj05) [inline rook/Tester.py:246]: I would like to suggest to use relative path to gold directory. @joshua-cogliati-inl Any suggestion here?

**Review by wangcj05 — COMMENTED** [review]
I have some minor comments for your consideration @dgarrett622. Hi @joshua-cogliati-inl , could you take a look at this PR?

File: rook/Tester.py (hunk 2 above)
Comment (dgarrett622) [reply rook/Tester.py:181]: Currently the path is relative to the test directory, not the gold directory. I initially wanted to have things relative to the gold directory, but some tests generate some data for a ROM, build the ROM, and then check that the ROM gives the same data generated previously. It would not be too difficult to change it to assume the path is relative to the gold directory.

File: rook/Tester.py (hunk 3 above)
Comment (dgarrett622) [reply rook/Tester.py:246]: There are a few tests that would need to have gold files moved to the gold directory to make this happen. I think it is probably better to do that and make a minor change here so that all gold files are in the gold directory.

File: tests/framework/PostProcessors/LimitSurface/tests [hunk 4]
```diff
@@ -27,14 +27,18 @@
...
  [../]
 
- [./testLimitSurfaceIntegralPPWihtBoundingError]
+ [./testLimitSurfaceIntegralPPWithBoundingError]
   type = 'RavenFramework'
   input = 'test_LimitSurface_with_err_bounds.xml'
   output = 'limitSurfaceBoundingError/LimitSurfaceWeightedPb_dump.xml'
-  csv = 'limitSurfaceBoundingError/LimitSurfaceWeightedPb_dump.csv'
   max_time = 300
   rel_err = 0.001
```
Comment (joshua-cogliati-inl) [inline tests/framework/PostProcessors/LimitSurface/tests:35]: The real test of this is if we can make the rel_err smaller. (That said, that can definitely be a separate pull request.)

File: rook/Tester.py (hunk 3 above)
Comment (joshua-cogliati-inl) [reply rook/Tester.py:246]: I am fine with it being relative to the test directory.

**Review by joshua-cogliati-inl — COMMENTED** [review]
I took a quick look, and I have no strong objections.  (I would prefer if it was not restricted to windows, mac, linux (see os_max_time))

**Review by joshua-cogliati-inl — DISMISSED** [review]

File: rook/Tester.py (hunk 3 above)
Comment (wangcj05) [reply rook/Tester.py:246]: If possible, I would like to have a consistent assumption about the path for the gold files.

Comment (dgarrett622) [reply rook/Tester.py:246]: > If possible, I would like to have a consistent assumption about the path for the gold files.

Okay, I will change them to be relative to the "gold" directory. I may need to move some gold files around or modify tests to place files in the "gold" directory, but it will all be consistent.

Comment (joshua-cogliati-inl) [reply rook/Tester.py]: Hm, changing them to be relative to the "gold" directory breaks the HERON plugin tests, so they will need to be fixed.

Comment (joshua-cogliati-inl) [reply rook/Tester.py]: (Sorry, but changes seem to propagate :( )

Comment (PaulTalbot-INL) [reply rook/Tester.py]: Correct me if I'm wrong, but it seems to me that we only assume relative to `gold` if you do NOT specify a custom path to the gold file; however, if you assign a custom path, it assumes relative to the test directory. Is that right?

Comment (dgarrett622) [reply rook/Tester.py]: @PaulTalbot-INL, yes, that is how it was set up originally. My most recent commit changes this to assume all given paths are relative to `gold`. This change did not affect most tests. The ARMA reseed precursor test required some "creative" path manipulations to get it to work where all gold files are relative to `gold`, but now all RAVEN tests pass. It looks like there is one HERON test that needs to be modified, but other plugins seem to be unaffected by this change.

File: tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml [hunk 5]
```diff
@@ -21,14 +21,14 @@
   </TestInfo>
 
   <RunInfo>
-    <WorkingDir>ARMAreseed</WorkingDir>
+    <WorkingDir>gold/ARMAreseed</WorkingDir>
```
Comment (wangcj05) [inline tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml:24]: why the working dir need to be within the gold the folder?

Comment (wangcj05) [reply tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml:24]: Could you explain it? I do not understand the test structure here.

**Review by wangcj05 — COMMENTED** [review]
I have some comments, could you take a look? @dgarrett622

File: tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml (hunk 5 above)
Comment (dgarrett622) [reply tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml:24]: It is a little strange. The `arma_reseed_test.xml` test unpickles an ARMA ROM and checks that the histories it generates are identical. The ARMA ROM and gold files for the `arma_reseed_test.xml` are generated by the `arma_reseed_precursor.xml` test. Since we want all gold files to be relative to the gold directory, I had to make `arma_reseed_precursor.xml` place the files in the gold directory so that the `arma_reseed_test.xml` test could find them correctly.

Comment (wangcj05) [reply tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml:24]: I see, I guess you can also use relative path for "arma_reseed_test.xml" test. The current test setup is a little strange for me. If we keep "arma_reseed_precursor.xml" without changes, and use relative path for "arma_reseed_test.xml"  in "tests", does it look better? I will let you decide which way want.

Comment (dgarrett622) [reply tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml:24]: Yeah, I don't like writing gold directory on the fly. I'll change it to use relative paths for `arma_reseed_test.xml`.

Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml]: Just to confirm, the `arma_reseed_test` should use the ARMA generated by the `arma_reseed_precursor` and not the one stored in `gold`, right?

Comment (dgarrett622) [reply tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml]: @PaulTalbot-INL yes, the `arma_reseed_test` uses the ARMA generated by the `arma_reseed_precursor` test. Before the most recent commit, this ARMA was generated and placed in the `gold` directory. Now it is generated and placed in the `ARMAreseed` directory not the `gold` directory.

**Review by wangcj05 — APPROVED** [review]
changes look good.

**Comment by wangcj05** [comment]
Heron plugin failed due to the changes in this PR. After the merge, @dgarrett622 please try to update HERON with the new changes. Thanks a lot for your contribution. Once the tests are green. PR can be merged.

**Comment by dgarrett622** [comment]
> Heron plugin failed due to the changes in this PR. After the merge, @dgarrett622 please try to update HERON with the new changes. Thanks a lot for your contribution. Once the tests are green. PR can be merged.

I'll put together a PR for the HERON test as soon as this gets merged in. Thanks!

**Comment by wangcj05** [comment]
Checklist is good. Thanks for your contribution! @dgarrett622 
In addition, please remember to update HERON.
